dependent events: followup

bnjbvr opened the pull request:
I'm tired. Here are some improvements/bugfixes for the dependent event mechanism.

andybalaam commented on crates/matrix-sdk-base/src/store/traits.rs:
```diff
@@ -1254,6 +1255,38 @@ pub enum DependentQueuedEventKind {
+/// A transaction id identifying a [`DependentQueuedEvent`] rather than its
+/// parent [`QueuedEvent`].
+///
+/// This thins wrapper adds some safety to some APIs, making it possible to
```
thins -> thin

andybalaam commented on crates/matrix-sdk/src/send_queue.rs:
```diff
@@ -1145,24 +1145,41 @@ impl SendHandle {
+
         match &d.kind {
-            DependentQueuedEventKind::Edit { .. } => latest_edit = Some(d),
+            DependentQueuedEventKind::Edit { .. } => {
```
Maybe a comment here like

`// Replace any previous edit with this one`

andybalaam commented on crates/matrix-sdk/src/send_queue.rs:
```diff
@@ -1145,24 +1145,41 @@ impl SendHandle {
 /// meaningful, i.e. the ones that wouldn't be overridden after applying the
 /// others.
 fn canonicalize_dependent_events(dependent: &[DependentQueuedEvent]) -> Vec<DependentQueuedEvent> {
```
So I think previously we assumed everything in `dependent` would be for the same parent and now we don't? Is this fixing an existing bug?

bnjbvr replied:
Yes indeed; this function was already handling all the dependent events for all the queued events, so this was a bug. The new test was failing on main.

andybalaam replied:
I assume there was not a github issue for it though? It feels weird to fix a real thing and have no audit trail, but I guess creating a whole issue just to close it seems a bit heavy.

bnjbvr replied:
No github issue, I've found out about the issues while addressing review comments in the PR that introduced the feature, so I figured I could address those in a followup, instead of sneaking in the changes in the approved PR.

andybalaam commented on crates/matrix-sdk/src/send_queue.rs:
```diff
@@ -924,6 +924,20 @@ impl QueueStorage {
 
         let canonicalized_dependent_events = canonicalize_dependent_events(&dependent_events);
+
+        // Get rid of the all non-canonical dependent events.
```
Does this fix an existing bug?

bnjbvr replied:
Yes indeed; didn't add a test for it, because it's a bit far fetched and shouldn't happen in theory that we don't have canonicalized events, and would require lots of setup etc. so little value vs what it's testing.

andybalaam replied:
No, happy to follow your instincts.

andybalaam reviewed (approved):
Looks good, thank you; some questions.

bnjbvr commented:
@andybalaam For what it's worth, I've added a small commit that's a big "duh" from me: the previous PR introduced a transaction id for the dependent event, but forgot to make use of it. This isn't a big deal for edits/redacts, but it would have been a massive failure for reactions. I've addressed it here for edits/redacts, which isn't a big change, but wanted to let you know before I autosquash and merge later.

andybalaam commented:
All looks good to me, thanks!
